Release version 4.0.0

mathieuboudreau opened the pull request:
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer
- [x] I've consulted [ADS's internal developer documentation](https://github.com/neuropoly/axondeepseg/wiki/Guidelines-for-contribution) to ensure my contribution is in line with any relevant design decisions


<!--- For the title, please observe the following rules:
	- Provide a concise and self-descriptive title
	- Do not include the applicable issue number in the title, do it in the PR body
	- If the PR is not ready for review, convert it to a draft.
-->

## Description
<!-- describe what the PR is about. Explain the approach and possible drawbacks.It's ok to repeat some text from the related issue. -->

Release version 4.0.0 of AxonDeepSeg that integrated IVADOMED into the project and provides Mac M1 compatibility.

## Linked issues
<!-- If the PR fixes any issues, indicate it here with issue-closing keywords: e.g. Resolves #XX, Fixes #XX, Addresses #XX. Note that if you want multiple issues to be autoclosed on PR merge, you must use the issue-closing verb before each relevant issue: e.g. Resolves #1, Resolves #2 -->

Resolves #523 #536

mathieuboudreau commented:
@mariehbourget this branch is ready for review.

mariehbourget commented:
Thank you @mathieuboudreau, I should be able to review by the end of the week.

Quick notes:
- We will need to update the documentation for v4 in this PR.
- We should probably do a "last" v3 release before the merge as well.

mathieuboudreau commented:
Looks like the cause was that we removed the tests for visualize.py, but we didn't remove the source file itself when we were doing cleanup in 3c903dc. I just deleted it

mathieuboudreau commented:
Hmm, not sure why the deleted file is still being counted in the number of lines during Coveralls

mariehbourget commented:
> Hmm, not sure why the deleted file is still being counted in the number of lines during Coveralls

Looks good now with +5% 🚀

mariehbourget commented:
The training [microscopy tutorial](https://ivadomed.org/tutorials/two_class_microscopy_seg_2d_unet.html) in ivadomed is now merged if we want to link to it in the documentation.

mathieuboudreau commented:
When we're ready to merge this branch, I'll update the checks settings from Python 3.7 to 3.8 to make them pass (we moved to 3.8 in this PR).

mathieuboudreau commented:
> * We should probably do a "last" v3 release before the merge as well.

Being done in #588

mathieuboudreau commented:
> * We will need to update the documentation for v4 in this PR.

Done in 61c59a3

mariehbourget commented on AxonDeepSeg/segment.py:
```diff
@@ -258,17 +196,14 @@ def main(argv=None):
+    ap.add_argument("-m", "--model", required=False, help='Folder where the model is located.')
     ap.add_argument('-s', '--sizepixel', required=False, help='Pixel size of the image(s) to segment, in micrometers. \n'+
                                                               'If no pixel size is specified, a pixel_size_in_micrometer.txt \n'+
                                                               'file needs to be added to the image folder path. The pixel size \n'+
```
For the "verbose" option at line 212 below. I think that verbose 2 and 3 are not used anymore in the code.
ivadomed is pretty verbose itself...
The only difference between verbose 0 and 1 is that with 1 it prints `Image /home/mhbourget/code/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_sem/data_test/image.png segmented.` after each segmented image.
I would be in favor of removing this option and default the verbose 1 behavior everywhere. What do you think?

mathieuboudreau replied:
Hmm I'm indifferent. I like having the option available for the future. What if we set it to 1 as default, and just say 0 is quiet mode and 1 may print out additional information useful for developers ?

mariehbourget replied:
The descriptions of 0 and 1 are OK as is, I was just suggesting 1 because there is only 1 line of difference between 0 and 1. We can keep both if you prefer, that's not a big issue.
However, I would remove 2 and 3 as they are not used at all anymore.

mathieuboudreau replied:
Resolved in 0438ad28f5f4a1d

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -231,17 +214,17 @@ The script to launch is called **axondeepseg**. It takes several arguments:
+                    The default BF model path is **model_seg_rat_axon-myelin_bf**.
 
 -s SIZEPIXEL        Pixel size of the image(s) to segment, in micrometers. 
                     If no pixel size is specified, a **pixel_size_in_micrometer.txt** file needs to be added to the image folder path ( that file should contain a single float number corresponding to the resolution of the image, i.e. the pixel size). The pixel size in that file will be used for the segmentation.
```
Verbose description at line 232 below to adjust to new behavior.

mathieuboudreau replied:
Resolved in c9eef2098

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -503,19 +486,13 @@ Jupyter notebooks
+* `02-morphometrics_extraction.ipynb <https://github.com/neuropoly/axondeepseg/blob/master/notebooks/04-morphometrics_extraction.ipynb>`_:
     Notebook that shows how to extract morphometrics from a sample segmented with AxonDeepSeg. The user can extract and save morphometrics for each axon (diameter, solidity, ellipticity, centroid, ...), estimate aggregate morphometrics of the sample from the axon/myelin segmentation (g-ratio, AVF, MVF, myelin thickness, axon density, ...), and generate overlays of axon/myelin segmentation masks,  ...
 
 .. NOTE ::
```
jupyter is now in the environment file, I don't think the `pip install jupyter` at line 501 below is necessary anymore.

mariehbourget replied:
For the warning at line 509.
The patch size of our new models is currently:
SEM: 256x256
TEM 512x512
BF: 512x512

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -536,56 +513,6 @@ Here is a list of useful Jupyter notebooks available with AxonDeepSeg:
 
    **Option:** If your image to segment is too small, you can use padding to artificially increase its size (i.e. add empty pixels around the borders).
 
-Guide for manual labelling
```
Oh! we will need a plan for this part.

So, now that we do the training in ivadomed:
- shouldn't we mention this somewhere and link to the tutorial?
- people will still need a "guide" for manual labelling, so the information here should be adjusted to the need of ivadomed for training, the question is, here or on ivadomed side?
- the differences are:
    - ivadomed use 2 separate black and white masks
    - ivaodmed uses BIDS structure (I'm a little bit stuck for documenting this part at the moment because the BEP is not merged yet).

I'll think about it!

mathieuboudreau replied:
I deleted it, because I assumed AxonDeepSeg doesn't deal with training at all anymore. I could see a simple link to an IVADOMED tutorial, but am unsure if we should include it in our own documentation as it might get out of sync quickly with IVADOMED.

Your call!

mathieuboudreau replied:
Maybe @jcohenadad can chime in too with his opinion on this

mariehbourget reviewed (commented):
Thank you @mathieuboudreau!

Here is the first part of my review, some comments/questions/corrections, but so far everything is working as expected.
I tested the installation (CPU), integration tests and basic segmentation CLI. I also reviewed the documentation.

General question about the documentation: how picky are we about the figures in the home page [here](https://axondeepseg.readthedocs.io/en/latest/)?
Those are from the original paper/architecture. With ivadomed, the "pipeline" is more or less the same but the network architecture is different (number of layers and numbers of features).  
We will most likely retrain the models in the upcoming months with maybe some changes so I would prefer not to change the figure for now but wanted to mention it anyway to have others' opinion.

I won't be able to finish the review today but will continue next week with the testing of morphometrics, notebooks and fsleyes.

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -495,19 +490,8 @@ Here is a list of useful Jupyter notebooks available with AxonDeepSeg:
-
 .. WARNING ::
-   The current models available for segmentation are trained for patches of 512x512 pixels. This means that your input image(s) should be at least 512x512 pixels in size **after the resampling to the target pixel size of the model you are using to segment**. 
+   The current models available for segmentation are trained for patches of 256x256 pixels for SEM and 512x512 pixels for TEM and BF. This means that your input image(s) should be at least 512x512 pixels in size **after the resampling to the target pixel size of the model you are using to segment**. 
```
```suggestion
   The current models available for segmentation are trained for patches of 256x256 pixels for SEM and 512x512 pixels for TEM and BF. This means that your input image(s) should be at least 256x256 or 512x512 pixels in size **after the resampling to the target pixel size of the model you are using to segment**. 
```

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -495,19 +490,8 @@ Here is a list of useful Jupyter notebooks available with AxonDeepSeg:
-
-        pip install jupyter
-
-    Then, go to the notebooks/ subfolder of AxonDeepSeg and launch a particular notebook as follows::
-
-        cd notebooks
-        jupyter notebook name_of_the_notebook.ipynb 
```
We could keep this part to launch the notebook and only delete the "pip install" part.

mathieuboudreau replied:
Done in 55795b02

mathieuboudreau commented on docs/source/documentation.rst:
```diff
@@ -231,17 +214,17 @@ The script to launch is called **axondeepseg**. It takes several arguments:
+                    The default BF model path is **model_seg_rat_axon-myelin_bf**.
 
 -s SIZEPIXEL        Pixel size of the image(s) to segment, in micrometers. 
                     If no pixel size is specified, a **pixel_size_in_micrometer.txt** file needs to be added to the image folder path ( that file should contain a single float number corresponding to the resolution of the image, i.e. the pixel size). The pixel size in that file will be used for the segmentation.
```
Done in 53aaa05

mathieuboudreau commented on docs/source/documentation.rst:
```diff
@@ -169,23 +169,6 @@ Known issues
 1. The FSLeyes installation doesn't always work on Linux. Refer to the `FSLeyes installation guide <https://users.fmrib.ox.ac.uk/~paulmc/fsleyes/userdoc/latest/install.html>`_ if you need. In our testing, most issues came from the installation of the wxPython package.
 
 
-GPU-compatible installation
```
That's done on the IVADOMED installation side right? Could we just link to the instructions on the IVADOMED page, in order to centralize the instructions since it may change in the future? https://ivadomed.org/installation.html#step-3-install-torch-and-torchvision-with-cpu-or-gpu-support

mariehbourget replied:
Yes, I was just unsure on how to proceed because we have our own torch requirments in our environment file [here](https://github.com/neuropoly/axondeepseg/blob/57b658300974053615914410a767327c7a03e0a9/environment.yml#L31).
This is working on CPU only.

For GPU, I tested on rosenberg like this:
1) install ADS with the conda environment file according to our instructions
2) I pip install afterwards the GPU requirements from [here](https://ivadomed.org/installation.html#step-3-install-torch-and-torchvision-with-cpu-or-gpu-support), from the `Nvidia GPU Support` and `All OS` tabs. (currently `pip install torch==1.8.1+cu111 torchvision==0.9.1+cu111 --find-links https://download.pytorch.org/whl/torch_stable.html`)
It works well, the segmentation uses the GPU and all the tests passed.

I think my broader question is:
Should we have the torch requirements in our environment file even for CPU or always link to the ivadomed doc to always be up-to-date? We may want to check how they do it in SCT.

mathieuboudreau replied:
Resolved

mariehbourget commented on notebooks/00-getting_started.ipynb:
```diff
@@ -183,9 +161,9 @@
     "**2.2. Launch the segmentation of the image.**\n",
     "\n",
-    "Here, we launch the segmentation. Here, we specify the following inputs in the *axon_segmentation* function: (i) the path of the image, (ii) the name of the image, (iii) the path of the model, (iv) the configuration json file of the model (v) the target resolution (resampled_resolutions) and (vi) the verbosity level. The target resolution of the current version of the models are 0.1 for the  ...
+    "Here, we launch the segmentation. Here, we specify the following inputs in the *axon_segmentation* function: (i) the path of the image, (ii) the name of the image, (iii) the path of the model, (iv) the configuration json file of the model (v) the target resolution (resampled_resolutions) and (vi) the verbosity level. The target resolution of the current version of the models are 0.1 for the  ...
```
```suggestion
    "Here, we launch the segmentation. Here, we specify the following inputs in the *segment_image* function: (i) the path of the image, (ii) the path of the model, (iii) the overlap value between patches and (iv) the resolution of the image (pixel size in micrometers).\n",
```

mariehbourget commented on notebooks/00-getting_started.ipynb:
```diff
@@ -98,7 +100,7 @@
     "**1.2. Select the trained model you want to use for the segmentation.**\n",
     "\n",
-    "Here, we specify the deep learning model we want to use in order to segment our sample. We currently propose 3 models: one for scanning electron microscopy (SEM) samples, a second for transmission electron microscopy (TEM) samples, and a third for bright field optical microscopy (OM) samples. The current versions of the models are **'default_SEM_model'**, **'default_TEM_model'**, and **'OM_m ...
+    "Here, we specify the deep learning model we want to use in order to segment our sample. We currently propose 3 models: one for scanning electron microscopy (SEM) samples, a second for transmission electron microscopy (TEM) samples, and a third for bright field optical microscopy (OM) samples. The current versions of the models are **'model_seg_rat_axon-myelin_sem'**, **'model_seg_mouse_axon- ...
```
```suggestion
    "Here, we specify the deep learning model we want to use in order to segment our sample. We currently propose 3 models: one for scanning electron microscopy (SEM) samples, a second for transmission electron microscopy (TEM) samples, and a third for bright field optical microscopy (BF) samples. The current versions of the models are **'model_seg_rat_axon-myelin_sem'**, **'model_seg_mouse_axon-myelin_tem\"'**, and **'model_seg_rat_axon-myelin_bf'**, respectively. In this case, our test sample is a SEM spinal cord sample of the rat, so we select the SEM model available."
```

mariehbourget commented on notebooks/00-getting_started.ipynb:
```diff
@@ -194,7 +172,12 @@
+    "        path_testing_image=path_testing / 'image.png',\n",
+    "        path_model=path_model,\n",
+    "        overlap_value=[48, 48],\n",
+    "        acquired_resolution=0.13\n",
```
The previous version with the axon_segmentation function was able to extract the pixel size from the txt file [here](https://github.com/neuropoly/axondeepseg/blob/45e7da020fd7b1e51ba84089252536a3619d98a4/AxonDeepSeg/apply_model.py#L238).
The new version of axon_segmentation does not have that feature so we need the acquired_resolution here. Would it be possible to bring back the feature in axon_segmentation so the resolution is extracted from the txt file instead?

mathieuboudreau replied:
Resolved in 0438ad28f5f4a1d

mariehbourget replied:
I tested it by removing the "acquired_resolution=0.13" in the "segment_image" call but I got an error, I tried a couple of thing but don't understand what is not working:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [7], in <module>
----> 1 segment_image(
      2         path_testing_image=path_testing / 'image.png',
      3         path_model=path_model,
      4         overlap_value=[48, 48])

File ~/code/axondeepseg/AxonDeepSeg/segment.py:112, in segment_image(path_testing_image, path_model, overlap_value, acquired_resolution, verbosity_level)
    109 img_name_original = acquisition_name.stem
    111 # Performing the segmentation
--> 112 axon_segmentation(path_acquisitions_folders=path_acquisition, acquisitions_filenames=[str(path_acquisition / acquisition_name)],
    113                   path_model_folder=path_model, overlap_value=overlap_value,
    114                   acquired_resolution=acquired_resolution)
    116 if verbosity_level >= 1:
    117     print(("Image {0} segmented.".format(path_testing_image)))

File ~/code/axondeepseg/AxonDeepSeg/apply_model.py:35, in axon_segmentation(path_acquisitions_folders, acquisitions_filenames, path_model_folder, overlap_value, acquired_resolution, verbosity_level)
     33 if (path_acquisitions_folders / 'pixel_size_in_micrometer.txt').exists():
     34     resolutions_file = open(path_acquisitions_folders / 'pixel_size_in_micrometer.txt', 'r')
---> 35     str_resolution = [float(file_.read()) for file_ in resolutions_file]
     36     acquired_resolution = float(str_resolution[0])
     37 else:

File ~/code/axondeepseg/AxonDeepSeg/apply_model.py:35, in <listcomp>(.0)
     33 if (path_acquisitions_folders / 'pixel_size_in_micrometer.txt').exists():
     34     resolutions_file = open(path_acquisitions_folders / 'pixel_size_in_micrometer.txt', 'r')
---> 35     str_resolution = [float(file_.read()) for file_ in resolutions_file]
     36     acquired_resolution = float(str_resolution[0])
     37 else:

AttributeError: 'str' object has no attribute 'read'
```

mariehbourget commented on AxonDeepSeg/apply_model.py:
```diff
@@ -1,493 +1,46 @@
-    # Generating the patch to acquisitions and loading the acquisitions resolutions.
-    path_acquisitions = [path_acquisitions_folders[i] / e for i, e in enumerate(acquisitions_filenames)]
-
-    # If we did not receive any resolution we read the pixel size in micrometer from each pixel.
-    if acquired_resolution == None:
-        if (path_acquisitions_folders[0] / 'pixel_size_in_micrometer.txt').exists():
-            resolutions_files = [open(path_acquisition_folder / 'pixel_size_in_micrometer.txt', 'r')
-                                 for path_acquisition_folder in path_acquisitions_folders]
-            acquisitions_resolutions = [float(file_.read()) for file_ in resolutions_files]
-        else:
-            exception_msg = "ERROR: No pixel size is provided, and there is no pixel_size_in_micrometer.txt file in image folder. " \
-                            "Please provide a pixel size (using argument -s), or add a pixel_size_in_micrometer.txt file " \
-                            "containing the pixel size value."
-            raise Exception(exception_msg)
```
I think this part should stay in the axon_segmentation function so we can extract the pixel size from the txt file using the 00-notebook (I left a similar comment there to).

mathieuboudreau replied:
Done in e785e81, hope it works as intented.

mariehbourget commented on config.py:
```diff
@@ -2,11 +2,11 @@
+myelin_suffix = Path('_seg-myelin-manual_pred.png')         # myelin segmentation suffix file name
 index_suffix = Path('_index.png')         # image with the index of the axons
-axonmyelin_index_suffix = Path('_axonmyelin_index.png') # Colored axonmyelin segmentation + the index image
+axonmyelin_index_suffix = Path('_axonmyelin-manual_pred_index.png') # Colored axonmyelin segmentation + the index image
```
```suggestion
axonmyelin_index_suffix = Path('_seg_axonmyelin-manual_pred_index.png') # Colored axonmyelin segmentation + the index image
```
For the current naming convention in this PR, we would need to add `_seg` here for consistency.
I suggest to not do the change for now, because I'll have some other news about suffixes that could change our strategy.

mathieuboudreau replied:
Ok, on hold

mariehbourget commented on ads_plugin.py:
```diff
@@ -529,7 +524,7 @@ def on_save_segmentation_button(self, event):
 
 
-        image_name = myelin_mask_overlay.name[:-len("_seg-myelin")]
+        image_name = myelin_mask_overlay.name.split("-myelin")[0]
```
```suggestion
        image_name = myelin_mask_overlay.name.split("_seg")[0]
```
For the current naming convention in this PR, we have to change to split at `_seg` otherwise the name of the files would have duplicated `_seg_seg`.
Ex:
![fsleyes-wrong-filename](https://user-images.githubusercontent.com/54086142/150016019-056081eb-9e81-4863-8525-0f9d41de5319.png)
Overlay names in the GUI are OK:
![fsleyes_overlay_name](https://user-images.githubusercontent.com/54086142/150016046-d0ec624f-2562-439b-b94d-b750330e7052.png)


I suggest to not do the change for now, because I'll have some other news about suffixes that could change our strategy.

mathieuboudreau replied:
Ok, on hold

mariehbourget reviewed (commented):
@mathieuboudreau
Here is the second part of my review.
I re-reviewed the doc and tested the CLI (segmentation and morpho) + fsleyes + notebooks.

I still need to test the GPU on rosenberg.
I will also have some news regarding the suffixes for prediction filenames. Give me some time and I should be able to explain by the end of the day.

Sorry for the several parts but it's a pretty intimidating PR 😬 and it helps to break it down for the review.

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -503,89 +481,27 @@ Jupyter notebooks
 
```
I found an error on lines 414 and 425 while reviewing (unrelated to this PR).
On both line, the command should be `axondeepseg_morphometrics` instead of `axondeepseg`.

mathieuboudreau replied:
Done in 289a55c

mariehbourget commented:
I have some news regarding the suffixes used for prediction filenames in ivadomed.

Previously, ivadomed was using the name of the ground truth in its prediction filenames which was resulting in not so great suffixes like `_seg-axon-manual_pred` and `_seg-myelin-manual_pred`.
We changed the ivadomed behavior, the naming is now agnostic of the task or the model and use the index of the class number (ex: `_class-0_pred` and `_class-1_pred`).

This has an impact on our v4 implementation because we followed the ivadomed old naming convention.
I think we should revert those changes and go back to the previous ADS suffixes that were `_seg-axon`, `_seg-myelin` and `_seg-axonmyelin`.

We can choose the suffixes and pass them as "target_lst" [here](https://github.com/neuropoly/axondeepseg/blob/57b658300974053615914410a767327c7a03e0a9/AxonDeepSeg/apply_model.py#L38).
All our models have in order "axon" and "myelin" as class 0 and 1.

mathieuboudreau commented on docs/source/documentation.rst:
```diff
@@ -169,23 +167,6 @@ Known issues
 1. The FSLeyes installation doesn't always work on Linux. Refer to the `FSLeyes installation guide <https://users.fmrib.ox.ac.uk/~paulmc/fsleyes/userdoc/latest/install.html>`_ if you need. In our testing, most issues came from the installation of the wxPython package.
```
Done in f167b7f

mathieuboudreau commented:
> I have some news regarding the suffixes used for prediction filenames in ivadomed.
> 
> Previously, ivadomed was using the name of the ground truth in its prediction filenames which was resulting in not so great suffixes like `_seg-axon-manual_pred` and `_seg-myelin-manual_pred`. We changed the ivadomed behavior, the naming is now agnostic of the task or the model and use the index of the class number (ex: `_class-0_pred` and `_class-1_pred`), see discussion in issue [ivadomed/ivadomed#1006](https://github.com/ivadomed/ivadomed/issues/1006) and PR [ivadomed/ivadomed#1043](https://github.com/ivadomed/ivadomed/pull/1043).
> 
> This has an impact on our v4 implementation because we followed the ivadomed old naming convention. I think we should revert those changes and go back to the previous ADS suffixes that were `_seg-axon`, `_seg-myelin` and `_seg-axonmyelin`.
> 
> We can choose the suffixes and pass them as "target_lst" [here](https://github.com/neuropoly/axondeepseg/blob/57b658300974053615914410a767327c7a03e0a9/AxonDeepSeg/apply_model.py#L38). All our models have in order "axon" and "myelin" as class 0 and 1.

Ok, so I'll start a list of the changes needed to be made, as I'm not 100% sure I fully understand everything yet.

* [ ] Update the IVADOMED version in our environment file https://github.com/neuropoly/axondeepseg/blob/289a55cf7dc74a2102176fc7d70acf63fc851cbd/environment.yml#L35
  * To what version?
* [ ] Update the test filenames for the proper ivadomed outputs, such as [this one](https://github.com/axondeepseg/data-testing/blob/main/__test_files__/__test_demo_files__/image_seg-axon-manual_pred.png)  and [this one](https://github.com/axondeepseg/data-testing/blob/main/__test_files__/__test_demo_files__/image_seg-axonmyelin-manual_pred.png), and [this one](https://github.com/axondeepseg/data-testing/blob/main/__test_files__/__test_demo_files__/invalid_gratio_seg-axon-manual_pred.png).
  * @mariehbourget could you provide examples of what the new filenames should be here? 
* [ ]  Update the config filenames to the old ones, as mentioned in your message, here:  https://github.com/neuropoly/axondeepseg/blob/45e7da020fd7b1e51ba84089252536a3619d98a4/config.py#L5-L7
* [ ] (unsure) add another set of suffixes in the config or the code that introduces `_class-0_pred` and `_class-1_pred`?
* [ ] (unsure) Update the test conditions for the new nomenclature, like here https://github.com/neuropoly/axondeepseg/blob/45e7da020fd7b1e51ba84089252536a3619d98a4/test/test_segment.py#L64-L71, if needed.
* [ ] Update FSLeyes for the new nomenclature.

mariehbourget commented:
@mathieuboudreau, thanks for the TODO list regarding suffixes. I edited directly your comment with my answers.

mariehbourget commented:
Here is another TODO list regarding DOCUMENTATION.
I have a few questions from my review and it would help to centralize them to discuss them in meeting.

- [ ] how picky are we about the figures in the home page [here](https://axondeepseg.readthedocs.io/en/latest/)?
Those are from the original paper/architecture. With ivadomed, the "pipeline" is more or less the same but the network architecture is different (number of layers and numbers of features).  
We will most likely retrain the models in the upcoming months with maybe some changes so I would prefer not to change the figure for now but wanted to mention it anyway to have others' opinion.

- [ ] How to document CPU vs. GPU install, see [comment ](https://github.com/neuropoly/axondeepseg/pull/587#discussion_r787222716).

- [ ] What do we do about the manual labelling part, see this [comment](https://github.com/neuropoly/axondeepseg/pull/587#discussion_r785202836).

- [ ] Should we include a link to ivadomed's installation and tutorial pages for training new models?

mariehbourget commented:
I have one last point for my review.
I tried comparing the morphometrics results between v3 and v4 (using the same v4 segmentation).

I got the same results except for the `solidity` and `orientation` values which differs between versions.
Those are extracted via `skimage measure.regionprops`, so maybe the version we use changed?

Should we open a separate issue/PR for this?

I don't have time to investigate further before our meeting on Friday.
It would be great if someone could look into this, @hermancollin maybe?

I also noticed that NaN values in the excel file from the CLI are "empty cells" in the excel file from FSLeyes, is that normal? tagging @Stoyan-I-A

hermancollin commented:
> I don't have time to investigate further before our meeting on Friday. It would be great if someone could look into this, @hermancollin maybe?

I'll look into this.

> I also noticed that NaN values in the excel file from the CLI are "empty cells" in the excel file from FSLeyes, is that normal? tagging @Stoyan-I-A

I noticed the same behaviour with the FSLeyes plugin on v3. I hadn't noticed it earlier as I had no access to FSLeyes but I don't think this is normal.

mathieuboudreau commented:
> I have one last point for my review. I tried comparing the morphometrics results between v3 and v4 (using the same v4 segmentation).
> 
> I got the same results except for the `solidity` and `orientation` values which differs between versions. Those are extracted via `skimage measure.regionprops`, so maybe the version we use changed?
> 
> Should we open a separate issue/PR for this?

@hermancollin did so in #589 .

I checked the difference in the compute morphometrics file; I didn't touch the calls to either solidity or orientation.

One thing that did change though is which version of `skimage` we install:

<img width="1369" alt="Screen Shot 2022-01-19 at 9 55 53 PM" src="https://user-images.githubusercontent.com/1421029/150248613-d995a6f2-bb1b-4164-92e0-ac384dcf23b3.png">

So, it's possibly an issue or something they fixed on their end? We could check if they have anything in the release notes, or if some default parameters changed.

Stoyan-I-A commented:
> I also noticed that NaN values in the excel file from the CLI are "empty cells" in the excel file from FSLeyes, is that normal?

I'll check it out soon. What is suspect is that we forgot to update the function in `ads_plugin.py` that generates the morphometrics.

mathieuboudreau commented:
> So, it's possibly an issue or something they fixed on their end? We could check if they have anything in the release notes, or if some default parameters changed.

It could have to do with this error message we were receiving while we were using 0.14.2 (V3), as we are now in a version >0.16.


>/srv/conda/envs/notebook/lib/python3.7/site-packages/skimage/measure/_regionprops.py:260: UserWarning: regionprops and image moments (including moments, normalized moments, central moments, and inertia tensor) of 2D images will change from xy coordinates to rc coordinates in version 0.16.

mathieuboudreau commented:
Orientation is related to moments; solidity unsure.
<img width="683" alt="Screen Shot 2022-01-19 at 10 12 47 PM" src="https://user-images.githubusercontent.com/1421029/150255644-0b0ceb33-c76b-47ac-8652-55537fe2f089.png">
<img width="423" alt="Screen Shot 2022-01-19 at 10 12 58 PM" src="https://user-images.githubusercontent.com/1421029/150255671-c4e0138b-ad73-4cd6-838e-32c3e00854dd.png">

mathieuboudreau commented:
See here https://github.com/scikit-image/scikit-image/issues/3840

>regionprops will use row-column coordinates in 0.16. You can start
using them now with regionprops(..., coordinates='rc'). You can silence
warning messages, and retain the old behavior, with
regionprops(..., coordinates='xy'). However, that option will go away
in 0.16 and result in an error. This change has a number of consequences.
Specifically, the "orientation" region property will measure the
anticlockwise angle from a vertical line, i.e. from the vector (1, 0) in
row-column coordinates.

mathieuboudreau commented:
We might be able to retain the old behavious with `coordinates='xy'` when calling regionprops.

mathieuboudreau commented:
> We might be able to retain the old behaviour with `coordinates='xy'` when calling regionprops.

Hmm, when I try that, I get:

> ValueError: Values other than "rc" for the "coordinates" argument to skimage.measure.regionprops are no longer supported. You should update your code to use "rc" coordinates and stop using the "coordinates" argument, or use skimage version 0.15.x or earlier.

Seems they only supported it for version 0.15x.

I had to move to version 0.16.x for Mac ARM1 compatibility if I recall, so there doesn't appear to be a way to revert to the old implementation that would provide the same values for at least `orientation`, and maybe also `solidity`, to resolve issue #589. @hermancollin if you can think of anything, let us know.

mariehbourget commented:
@mathieuboudreau I prepared some [slides](https://docs.google.com/presentation/d/1wxBsS-X2fVDEc--sGLEHBXrXb27yjGw1d0IiMUIBqq8/edit#slide=id.gf8eba4e2b3_0_119) to explain the naming convention idea, I'll present it in meeting tomorrow.

mathieuboudreau commented:
Here is the error I get: 

```
Traceback (most recent call last):
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/bin/axondeepseg_test", line 33, in <module>
    sys.exit(load_entry_point('AxonDeepSeg', 'console_scripts', 'axondeepseg_test')())
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/integrity_test.py", line 35, in integrity_test
    axon_segmentation(path_testing, [str(path_testing / image)], path_model, overlap_value=[48,48], acquired_resolution=0.13)
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/apply_model.py", line 52, in axon_segmentation
    imed_inference.pred_to_png(nii_lst, target_lst, str(Path(input_filenames[0]).parent / Path(input_filenames[0]).stem))
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/inference.py", line 226, in pred_to_png
    filename = subj_path + target + "_pred" + suffix + ".png"
TypeError: can only concatenate str (not "list") to str
```

Actually, I think the error might be in my syntax, but it still seems to want to add "_pred" here if I read it correctly, no?

mathieuboudreau commented:
Closer: 

```
Traceback (most recent call last):
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/bin/axondeepseg_test", line 33, in <module>
    sys.exit(load_entry_point('AxonDeepSeg', 'console_scripts', 'axondeepseg_test')())
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/integrity_test.py", line 35, in integrity_test
    axon_segmentation(path_testing, [str(path_testing / image)], path_model, overlap_value=[48,48], acquired_resolution=0.13)
  File "/Users/mathieuboudreau/neuropoly/github/axondeepseg/AxonDeepSeg/apply_model.py", line 52, in axon_segmentation
    imed_inference.pred_to_png(nii_lst, target_lst, str(Path(input_filenames[0]).parent / Path(input_filenames[0]).stem))
  File "/Users/mathieuboudreau/opt/anaconda3/envs/ads_venv/lib/python3.8/site-packages/ivadomed/inference.py", line 226, in pred_to_png
    filename = subj_path + target + "_pred" + suffix + ".png"
TypeError: can only concatenate str (not "PosixPath") to str
```

Not sure which PosixPath I have. 

```python
 # IVADOMED automated segmentation
 nii_lst, _ = imed_inference.segment_volume(str(path_model), input_filenames, options=options)
    
 target_lst = [axon_suffix, myelin_suffix]

 imed_inference.pred_to_png(nii_lst, target_lst, str(Path(input_filenames[0]).parent / Path(input_filenames[0]).stem))
```

and

```python


# segmentation files suffix names
axonmyelin_suffix = Path('_seg-axonmyelin.png') # axon + myelin segmentation suffix file name
axon_suffix = Path('_seg-axon.png')             # axon segmentation suffix file name
myelin_suffix = Path('_seg-myelin.png')         # myelin segmentation suffix file name
index_suffix = Path('_index.png')         # image with the index of the axons
axonmyelin_index_suffix = Path('_axonmyelin_index.png') # Colored axonmyelin segmentation + the index image


```

mathieuboudreau commented:
Ahh, I see it now. Fixing it

mathieuboudreau commented:
Ok I have this now, but this isn't what you proposed, right @mariehbourget ?

<img width="272" alt="Screen Shot 2022-01-21 at 1 21 17 PM" src="https://user-images.githubusercontent.com/1421029/150571650-f114ae31-9f7e-4ea5-b969-b9b4df7196db.png">

See the mismatch between the output filenames and what I provided in the config:

```python
# segmentation files suffix names
axon_suffix = Path('_seg-axon.png')             # axon segmentation suffix file name
myelin_suffix = Path('_seg-myelin.png')         # myelin segmentation suffix file name
```

mariehbourget commented:
> Ok I have this now, but this isn't what you proposed, right @mariehbourget ?
> 
> <img alt="Screen Shot 2022-01-21 at 1 21 17 PM" width="272" src="https://user-images.githubusercontent.com/1421029/150571650-f114ae31-9f7e-4ea5-b969-b9b4df7196db.png">
> 
> See the mismatch between the output filenames and what I provided in the config:
> 
> ```python
> # segmentation files suffix names
> axon_suffix = Path('_seg-axon.png')             # axon segmentation suffix file name
> myelin_suffix = Path('_seg-myelin.png')         # myelin segmentation suffix file name
> ```

Thank you @mathieuboudreau for investigating.
Sorry, I totally missed the hardcoded "_pred" part when making my suggestion.

And you are right, I will need to fix it on ivadomed's side unless we choose to keep the "pred". I can take care of this early next week.

mariehbourget commented:
@mathieuboudreau, I opened a draft PR on ivaodmed https://github.com/ivadomed/ivadomed/pull/1050 ready to test.

mariehbourget commented:
@mathieuboudreau, I don't know if you was my previous message.
I opened a PR in ivadomed to fix the prediction filenames that is ready to test with your new implementation for prediction filenames.

I also wanted to link a new issue on SCT that will deal with the same kind of implementation for setting the prediction filenames for different models: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3662.
Currently, all our models are `class-0 = axon` and `class-1 = myelin`, so we don't need specific suffixes per model but it could happen in the future.

mariehbourget commented:
Hi @mathieuboudreau,
The fix for prediction filenames in ivadomed is merged and a new version 2.9.3 released.(https://github.com/ivadomed/ivadomed/releases/tag/v2.9.3)

mathieuboudreau commented:
@mariehbourget I think I have completed (or at least, drafted solutions for review all the review suggestions you've made. Please have another look and let me know what remains to be done prior to releasing this version!

mariehbourget reviewed (commented):
Thank you @mathieuboudreau!

I tested the installation on CPU and GPU and all the changes related to the naming convention.
For our checklists, I think everything is there.

I left a few comments for documentation formatting and the ads_plugin.
There is only one feature that does not work right now, I'm unable to read the pixel size from the pixel_in_micrometer.txt file when calling directly `segment_image` without `acquired_resolutions`, comment in notebook-00.

mathieuboudreau commented:
@mariehbourget I just accepted all suggestions, and fixed the bug in reading resolutions in commit d0abfaa50f72fa4aceb24994783f29c2ba229bfb

mariehbourget commented on CHANGELOG.md:
```diff
@@ -1,6 +1,17 @@
 Changelog
 ===============================================================================
 
+## Version 4.0.0 - 2022-01-14
```
Do you want to add changes from #595?

Should we add a note somewhere about the changes in morphometrics from #589?

mathieuboudreau replied:
Yes for 589, no for 595 (just tests). On it.

mathieuboudreau replied:
Done in [dea20b0](https://github.com/neuropoly/axondeepseg/pull/587/commits/dea20b01e494866e815295527385b4a028748390)

mariehbourget reviewed (approved):
Thank you @mathieuboudreau!

I think everything is working as expected.
I left some minor comment/suggestions but approve the PR since they are small details.

Question: I saw that #592 and #584 are approved, do we want to integrate them before merging?

mathieuboudreau commented:
> Question: I saw that #592 and #584 are approved, do we want to integrate them before merging?

Done, just merged them into this branch.

mathieuboudreau commented:
@mariehbourget I've merged the two PRs, accepted your suggestions, and updated the Changelog per your suggestion.

If this is ready to merge, please give it one final approval!

mariehbourget commented on CHANGELOG.md:
```diff
@@ -44,7 +51,7 @@ Changelog
  - Move test files to OSF, add download functionality [View pull request](https://github.com/neuropoly/axondeepseg/pull/373)
  - Add postprocessing tests [View pull request](https://github.com/neuropoly/axondeepseg/pull/365)
-
+ - Changed all instances of imageio_imread by ads_utils.imread [View pull request](https://github.com/neuropoly/axondeepseg/pull/584)
```
I think this should be in v4.0.0 and not v3.3.0

mathieuboudreau replied:
Woops, I miscopied early on and forgot to revert. Thanks!

mariehbourget commented on AxonDeepSeg/segment.py:
```diff
@@ -258,42 +196,36 @@ def main(argv=None):
                                                               'file needs to be added to the image folder path. The pixel size \n'+
                                                               'in that file will be used for the segmentation.',
                                                               default=None)
     ap.add_argument('-v', '--verbose', required=False, type=int, choices=list(range(0,4)), help='Verbosity level. \n'+
```
```suggestion
    ap.add_argument('-v', '--verbose', required=False, type=int, choices=list(range(0,2)), help='Verbosity level. \n'+
```

mariehbourget replied:
Hey @mathieuboudreau, I don't see this change, is it normal?

mariehbourget reviewed (commented):
Thanks @mathieuboudreau!
2 small things, I promise to approve after that! :)

mathieuboudreau commented:
> Thanks @mathieuboudreau!
> 2 small things, I promise to approve after that! :)

Perfect! Done and done =)

mathieuboudreau commented:
Sorry, I clicked commit changed but it didn't work the first time! It should be there now.

--
Mathieu Boudreau, PhD
Research Fellow | Montreal Heart Institute
Software Developer | qMRLab and AxonDeepSeg
Lead Editor | MRM Highlights<http://linkedin.com/company/mrm-highlights/><https://www.linkedin.com/company/mrm-highlights/>

________________________________
From: mariehbourget ***@***.***>
Sent: Thursday, February 3, 2022 13:44
To: neuropoly/axondeepseg ***@***.***>
Cc: Mathieu Boudreau ***@***.***>; Mention ***@***.***>
Subject: Re: [neuropoly/axondeepseg] Release version 4.0.0 (PR #587)


@mariehbourget commented on this pull request.

________________________________

In AxonDeepSeg/segment.py<https://github.com/neuropoly/axondeepseg/pull/587#discussion_r798818755>:

>      requiredName.add_argument('-i', '--imgpath', required=True, nargs='+', help='Path to the image to segment or path to the folder \n'+
                                                                                 'where the image(s) to segment is/are located.')

-    ap.add_argument("-m", "--model", required=False, help='Folder where the model is located. \n'+
-                                                          'The default SEM model path is: \n'+str(default_SEM_path)+'\n'+
-                                                          'The default TEM model path is: \n'+str(default_TEM_path)+'\n'+
-                                                          'The default OM model path is: \n'+str(model_seg_pns_bf_path)+'\n')
+    ap.add_argument("-m", "--model", required=False, help='Folder where the model is located, if different from the default model.')
     ap.add_argument('-s', '--sizepixel', required=False, help='Pixel size of the image(s) to segment, in micrometers. \n'+
                                                               'If no pixel size is specified, a pixel_size_in_micrometer.txt \n'+
                                                               'file needs to be added to the image folder path. The pixel size \n'+
                                                               'in that file will be used for the segmentation.',
                                                               default=None)
     ap.add_argument('-v', '--verbose', required=False, type=int, choices=list(range(0,4)), help='Verbosity level. \n'+


Hey @mathieuboudreau<https://github.com/mathieuboudreau>, I don't see this change, is it normal?

—
Reply to this email directly, view it on GitHub<https://github.com/neuropoly/axondeepseg/pull/587#discussion_r798818755>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAK25ZJEMZGCNPTNOBQUTODUZK5GFANCNFSM5LZGIOJQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

mariehbourget reviewed (approved):
Thank you so much @mathieuboudreau for coordinating this project. 🚀 
Version 4 is ready! 😀

mathieuboudreau commented:
> Thank you so much @mathieuboudreau for coordinating this project. 🚀
> Version 4 is ready! 😀

Thank *you* for your very rigourous review @mariehbourget and all your suggestsions !!
